packages/fxa-settings/src/pages/InlineRecoveryKeySetup/container.tsx

```diff
@@ -3,22 +3,147 @@
+  };
+  const { email, uid, sessionToken, unwrapBKey } = location.state || {};
+
+  const sensitiveDataClient = useSensitiveDataClient();
+  const sensitiveData = sensitiveDataClient.getData(AUTH_DATA_KEY);
+  const { authPW, emailForAuth } =
+    (sensitiveData as unknown as { emailForAuth: string; authPW: string }) ||
+    {};
+
+  const navigateForward = useCallback(() => {
+    console.log('currentStep:', currentStep);
+    setCurrentStep(currentStep + 1);
+  }, [currentStep]);
+
+  const createRecoveryKey = useCallback(
+    (
+        uid: string,
+        sessionToken: string,
+        unwrapBKey: string
+      ): (() => Promise<CreateRecoveryKeyHandler>) =>
+      async () => {
+        try {
+          // We must reauth for another `keyFetchToken` because we sent it to Sync
+          const reauth = await authClient.sessionReauthWithAuthPW(
+            sessionToken,
+            emailForAuth,
+            authPW,
+            {
+              keys: true,
+              reason: 'recovery_key',
+            }
+          );
+          // TODO, return error if reauth is not successful or keyFetchToken is not defined
+          const keys = await authClient.accountKeys(
+            reauth.keyFetchToken!,
+            unwrapBKey
+          );
+          const { recoveryKey, recoveryKeyId, recoveryData } =
+            await generateRecoveryKey(uid, keys);
+          await authClient.createRecoveryKey(
+            sessionToken,
+            recoveryKeyId,
+            recoveryData
+          );
+
+          cache.modify({
+            id: cache.identify({ __typename: 'Account' }),
+            fields: {
+              recoveryKey() {
+                return {
+                  exists: true,
+                };
+              },
+            },
+          });
+          // TODO, share base32Encode bits with one in settings
+          setFormattedRecoveryKey(
+            base32Encode(recoveryKey.buffer, 'Crockford')
+              .match(/.{4}/g)!
+              .join(' ')
+          );
+          navigateForward();
+          return { data: { recoveryKey } };
+        } catch (error) {
+          return getHandledError(error);
+        }
+      },
+    [authClient, emailForAuth, authPW, navigateForward]
+  );
+
+  const updateRecoveryHint = useCallback(
+    (sessionToken: string): ((hint: string) => Promise<void>) =>
+      async (hint: string) => {
+        // The try/catch for this is handled in the component where it's called
+        // because we're sharing the Hint component with Settings + Account.ts
+        authClient.updateRecoveryKeyHint(sessionToken, hint);
+      },
+    [authClient]
+  );
+
+  if (
+    !uid ||
+    !sessionToken ||
+    !unwrapBKey ||
+    !email ||
+    !emailForAuth ||
+    !authPW
+  ) {
+    // go to CAD with success messaging, we do not want to re-prompt for password
+    const { to } = getSyncNavigate(location.search);
+    hardNavigate(to);
+    return <></>;
+  }
+
+  // Curry already checked values
```

Comment:
There's a few places in other container components I think we could benefit from doing this in. This avoids having to pass these values into `InlineRecoveryKeySetup` and avoids having to tell typescript with `!` that these are not undefined.

Reply:
Very nice 🙌 Also simplifies a lot of the mocking for tests and stories